File: previous years/2022/2022-03-09-gray-shades.py
import numpy as np
import cv2 as cv
import time

img = np.zeros((512, 512), np.uint8)
print(img.shape)

t0 = time.time()
for y in range(0, img.shape[0]):
    for x in range(0, img.shape[1]):
        img[y][x] = (y / (img.shape[0]-1))*255
print(f"Elapsed time = {time.time()-t0:.3f}")


cv.imshow("Image", img)
cv.waitKey(0)
cv.destroyAllWindows()

File: previous years/2022/2022-03-09-gray-shades.cpp
// include aia and ucas utility functions
#include "aiaConfig.h"
#include "ucasConfig.h"

int main()
{
	cv::Mat img(512, 512, CV_8U, cv::Scalar(0));
	aia::imshow("Image", img);

	ucas::Timer timer;

	for (int y = 0; y < img.rows; y++)
	{
		for (int x = 0; x < img.cols; x++)
			img.at<unsigned char>(y, x) = float(y) / (img.rows - 1) * 255;
	}
	printf("Elapsed time (slow version) = %.3f\n", timer.elapsed<float>());

	timer.restart();
	for (int y = 0; y < img.rows; y++)
	{
		unsigned char* yRow = img.ptr<unsigned char>(y);
		for (int x = 0; x < img.cols; x++)
			yRow[x] = float(y) / (img.rows - 1) * 255;
	}

	printf("Elapsed time (fast version) = %.3f\n", timer.elapsed<float>());
	aia::imshow("Image", img);

	return EXIT_SUCCESS;
	
}

